035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+
+```yaml
+spec:
+  clusterCa:
+    validityDays: <integer>
+    generateCertificateAuthority: <boolean>
+    generateSecretOwnerReference: <boolean>
+    renewalDays: <integer>
+    certificateExpirationPolicy: <renew-certificate|replace-key>
+```
+
+We can use the usual `type` discriminator mechanism to support different kinds of issuer (i.e. the type would select the implementations of `CertificateIssuer`).
+
+#### `type: internal`
+
+If `type` was omitted the `internal` type would be used. The `internal` type supports all the properties given above. 
+The corresponding `InternalCertificateIssuer` would maintain backwards compatibility, using a root CA directly signing EE certificates.
+
+#### `type: cert-manager.io`
+
+A `cert-manager.io` type could be supported which would use `cert-manager.io/v1` [`CertificateRequest`][cmioCertificateRequest] to issue certificates.
+
+```yaml
+spec:
+  clusterCa:
+    type: cert-manager.io
+    validityDays: <integer> # used for CertificateRequest.spec.duration
+    issuer:
+      name: <string>  # used for CertificateRequest.spec.duration
+      kind: <string>
+      group: <string>
+    generateSecretOwnerReference: <boolean>
+    renewalDays: <integer>
+```
+
+This would result in the creation of a resource like the following:
+
+```yaml
+apiVersion: cert-manager.io/v1
+kind: CertificateRequest
+metadata:
+  name: <string> # provided by Strimzi operator
+spec:
+  request: <string> # PEM encoded CSR
+  isCA: false
+  usages: # provided by Strimzi operator
+  - signing
+  - digital signature
+  - server auth
+  duration: <integer> # from the Strimzi CR validityDays
+  issuerRef: # from the Strimzi CR issuer
+    name: ...
+    kind: ...
+    group: ...
+```
+
+Notes: 
+* The `spec.request`, being a CSR, could include the subject alternative names etc. for the required certificate.
+* The `spec.usages` would be implied by the kind of certificate being generated, e.g. broker certs would need `server auth`, but client certs issued by the UO would not.
+
+
+### `type: vault`
+
+Vault has [support for issuing certificates](https://www.vaultproject.io/docs/secrets/pki) 
+and it can be used via [a REST API](https://www.vaultproject.io/api/secret/pki#generate-certificate). 
+
+A `VaultCertificateIssuer` would be work similarly to the cert-manager type, but rather making the request using a Custom Resource, it would be implemented by using the REST API directly. 
+In Strimzi the CR might look like this:
+
+```yaml
+spec:
+  clusterCa:
+    type: vaultproject.io/pki
+    validityDays: <integer> # used for CertificateRequest.spec.duration
+    vaultUrl: <a URL to the vault REST API>
+    vaultToken:
+      secretName: <a Secret containing a vault token>
+    generateSecretOwnerReference: <boolean>
+    renewalDays: <integer>
+```
+
+When a certificate was issued the `VaultCertificateIssuer` would make 
+a POST request to `/pki/issue/:name` and extract the issued certificate from the response.
```

Comment:
So, what would we do with the certificates once we get them? Store them in secrets I assume based on the _non-goals_?  Maybe it should be specifically mentioned here.

Reply:
I'll mention in the text